core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -1771,6 +1776,141 @@ class KafkaController(val config: KafkaConfig,
+          case Some(leaderIsrAndControllerEpoch) =>
+            val currentLeaderAndIsr = leaderIsrAndControllerEpoch.leaderAndIsr
+            if (newLeaderAndIsr.leader != currentLeaderAndIsr.leader) {
+              Errors.NOT_LEADER_OR_FOLLOWER
+            } else if (newLeaderAndIsr.leaderEpoch < currentLeaderAndIsr.leaderEpoch) {
+              Errors.FENCED_LEADER_EPOCH
+            } else {
+              val currentAssignment = controllerContext.partitionReplicaAssignment(tp)
+              if (!newLeaderAndIsr.isr.forall(replicaId => currentAssignment.contains(replicaId))) {
+                warn(s"Some of the proposed ISR are not in the assignment for partition $tp. Proposed ISR=$newLeaderAndIsr.isr assignment=$currentAssignment")
+                Errors.INVALID_REQUEST
+              } else if (!newLeaderAndIsr.isr.forall(replicaId => controllerContext.isReplicaOnline(replicaId, tp))) {
+                warn(s"Some of the proposed ISR are offline for partition $tp. Proposed ISR=$newLeaderAndIsr.isr")
+                Errors.INVALID_REQUEST
+              } else {
+                Errors.NONE
+              }
+            }
+          case None => Errors.UNKNOWN_TOPIC_OR_PARTITION
+        }
+        if (partitionError == Errors.NONE) {
+          // Bump the leaderEpoch for partitions that we're going to write
+          Some(tp -> newLeaderAndIsr.newEpochAndZkVersion)
+        } else {
+          partitionErrors.put(tp, partitionError)
+          None
+        }
+    }
+
+    // Do the updates in ZK
+    info(s"Updating ISRs for partitions: ${adjustedIsrs.keySet}.")
+    val UpdateLeaderAndIsrResult(finishedUpdates, badVersionUpdates) =  zkClient.updateLeaderAndIsr(
+      adjustedIsrs, controllerContext.epoch, controllerContext.epochZkVersion)
+
+    val successfulUpdates: Map[TopicPartition, LeaderAndIsr] = finishedUpdates.flatMap {
+      case (partition: TopicPartition, isrOrError: Either[Throwable, LeaderAndIsr]) =>
+      isrOrError match {
+        case Right(updatedIsr) =>
+          info("ISR for partition %s updated to [%s] and zkVersion updated to [%d]".format(partition, updatedIsr.isr.mkString(","), updatedIsr.zkVersion))
+          Some(partition -> updatedIsr)
+        case Left(error) =>
+          warn(s"Failed to update ISR for partition $partition", error)
+          partitionErrors.put(partition, Errors.forException(error))
+          None
+      }
+    }
+
+    badVersionUpdates.foreach(partition => {
+      warn(s"Failed to update ISR for partition $partition, bad ZK version")
+      partitionErrors.put(partition, Errors.INVALID_ISR_VERSION)
+    })
+
+    // Update our cache
+    updateLeaderAndIsrCache(successfulUpdates.keys.toSeq)
+
+    // Send back AlterIsr response
+    callback.apply(Left(partitionErrors))
+
+    // Send out LeaderAndIsr for successful updates
+    brokerRequestBatch.newBatch()
+
+    // Send LeaderAndIsr for all requested partitions
```

Comment:
I think we need to send LeaderAndIsr for all the given partitions whether we updated the ISR or not. In cases where we failed due, the leaders likely have stale metadata. This way we can proactively send them the latest state.